File: scripts/media-gui.mjs
#!/usr/bin/env node

/**
 * Lightweight local GUI to manage portfolio media.
 * - Runs an Express server with a single-page UI.
 * - Lets you upload images/videos into static/ (carousel + portfolio).
 * - Writes entries into src/lib/data/portfolio-data.json.
 *
 * Local-only: there is no auth. Run when you need to update media, then stop it.
 *
 * Usage:
 *   pnpm add -D express multer   # once, if not installed
 *   npm run media-gui
 */

import express from 'express';
import multer from 'multer';
import fs from 'node:fs';
import path from 'node:path';

const root = process.cwd();
const port = Number(process.env.MEDIA_GUI_PORT || 4174);

const dataPath = path.join(root, 'src', 'lib', 'data', 'portfolio-data.json');
const imageRoot = path.join(root, 'static', 'images', 'portfolio');
const videoRoot = path.join(root, 'static', 'videos', 'portfolio');

const categoryList = [
	{ id: '3d-tours', name: '3D Tours' },
	{ id: 'interior', name: 'Interior Photography' },
	{ id: 'exterior', name: 'Exterior Photography' },
	{ id: 'drone', name: 'Drone & Aerial' },
	{ id: 'virtual-staging', name: 'Virtual Staging' },
	{ id: 'floorplans', name: 'Floorplans (Cubicasa)' },
	{ id: 'portraits', name: 'Portraits & Headshots' }
];

function ensureDefaults() {
	[imageRoot, videoRoot].forEach((dir) => fs.mkdirSync(dir, { recursive: true }));
	categoryList.forEach((c) => fs.mkdirSync(path.join(imageRoot, c.id), { recursive: true }));
	if (!fs.existsSync(dataPath)) {
		fs.writeFileSync(dataPath, '[]', 'utf8');
	}
}

function readData() {
	try {
		return JSON.parse(fs.readFileSync(dataPath, 'utf8'));
	} catch {
		return [];
	}
}

function writeData(items) {
	fs.writeFileSync(dataPath, JSON.stringify(items, null, 2));
}

const storage = multer.diskStorage({
	destination: (req, file, cb) => {
		const isVideo = file.mimetype.startsWith('video/');
		const category = req.body.category || 'uncategorized';
		const dest = isVideo ? videoRoot : path.join(imageRoot, category);
		fs.mkdirSync(dest, { recursive: true });
		cb(null, dest);
	},
	filename: (_req, file, cb) => {
		const ext = path.extname(file.originalname);
		const base = path.basename(file.originalname, ext);
		const safe =
			base
				.toLowerCase()
				.replace(/[^a-z0-9]+/g, '-')
				.replace(/^-+|-+$/g, '') || 'media';
		cb(null, `${Date.now()}-${safe}${ext}`);
	}
});

const upload = multer({ storage });

ensureDefaults();

const app = express();
app.use(express.json());
app.use(express.urlencoded({ extended: true }));

app.get('/api/data', (_req, res) => {
	const items = readData();
	res.json({ categories: categoryList, items });
});

app.post('/api/upload', upload.single('file'), (req, res) => {
	if (!req.file) {
		return res.status(400).json({ error: 'File is required' });
	}

	const items = readData();
	const nextId = items.reduce((max, item) => Math.max(max, Number(item.id) || 0), 0) + 1;

	const isVideo = req.file.mimetype.startsWith('video/');
	const category = req.body.category || 'interior';
	const publicPath = isVideo
		? `/videos/portfolio/${req.file.filename}`
		: `/images/portfolio/${category}/${req.file.filename}`;

	const tags =
		(req.body.tags || '')
			.split(',')
			.map((t) => t.trim())
			.filter(Boolean) || [];

	const newItem = {
		id: nextId,
		title: req.body.title || req.file.originalname,
		category,
		description: req.body.description || '',
		image: isVideo ? undefined : publicPath,
		video: isVideo ? publicPath : undefined,
		tags,
		date: req.body.date || new Date().toISOString().slice(0, 10),
		client: req.body.client || '',
		results: req.body.results || ''
	};

	items.push(newItem);
	writeData(items);

	res.json({ ok: true, item: newItem });
});

app.get('/', (_req, res) => {
	res.type('html').send(renderPage());
});

app.listen(port, () => {
	console.log(`Media GUI running at http://localhost:${port}`);
	console.log('Tip: stop the server when you are done (Ctrl+C).');
});

function renderPage() {
	return `<!doctype html>
<html lang="en">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <title>Full Scope Media – Media GUI</title>
  <style>
    :root { font-family: 'Inter', system-ui, -apple-system, sans-serif; color: #0f172a; background: #f8fafc; }
    body { margin: 0; padding: 24px; }
    h1 { margin: 0 0 16px; }
    .layout { display: grid; grid-template-columns: 420px 1fr; gap: 24px; align-items: start; }
    form { background: #fff; border: 1px solid #e2e8f0; border-radius: 12px; padding: 16px; box-shadow: 0 10px 30px rgba(15,23,42,0.06); }
    label { display: block; font-weight: 600; margin-top: 12px; }
    input, select, textarea { width: 100%; padding: 10px 12px; border: 1px solid #cbd5e1; border-radius: 8px; font-size: 14px; }
    textarea { min-height: 80px; resize: vertical; }
    button { margin-top: 16px; padding: 12px 14px; background: linear-gradient(135deg, #111827, #0f172a); color: #fff; border: none; border-radius: 10px; font-weight: 700; cursor: pointer; }
    button:hover { filter: brightness(1.05); }
    .pill { display: inline-flex; align-items: center; gap: 6px; background: #e2e8f0; padding: 6px 10px; border-radius: 999px; font-size: 12px; margin: 0 6px 6px 0; }
    .grid { display: grid; grid-template-columns: repeat(auto-fill, minmax(240px, 1fr)); gap: 16px; }
    .card { background: #fff; border: 1px solid #e2e8f0; border-radius: 12px; overflow: hidden; box-shadow: 0 10px 30px rgba(15,23,42,0.06); }
    .card img, .card video { width: 100%; display: block; aspect-ratio: 4 / 3; object-fit: cover; background: #0f172a10; }
    .card .content { padding: 12px; }
    .muted { color: #475569; font-size: 13px; }
    .tags { margin-top: 8px; }
    .banner { margin-bottom: 16px; padding: 12px 16px; background: #111827; color: #f8fafc; border-radius: 12px; }
    @media (max-width: 900px) { .layout { grid-template-columns: 1fr; } }
  </style>
</head>
<body>
  <h1>Full Scope Media — Media Manager</h1>
  <div class="banner">
    <strong>Local-only.</strong> Uploads save into <code>static/</code> and update <code>portfolio-data.json</code>. Stop this server when finished.
  </div>
  <div class="layout">
    <form id="upload-form">
      <label>Category
        <select name="category" required id="category"></select>
      </label>
      <label>Title
        <input name="title" placeholder="Kitchen refresh in Okemos" />
      </label>
      <label>Description
        <textarea name="description" placeholder="Short description"></textarea>
      </label>
      <label>Client (optional)
        <input name="client" placeholder="Acme Realty" />
      </label>
      <label>Results (optional)
        <input name="results" placeholder="e.g., Listing under contract in 5 days" />
      </label>
      <label>Date (YYYY-MM-DD)
        <input name="date" type="date" />
      </label>
      <label>Tags (comma-separated)
        <input name="tags" placeholder="interior, natural light, drone" />
      </label>
      <label>Media file (image or video)
        <input type="file" name="file" accept="image/*,video/*" required />
      </label>
      <button type="submit">Upload &amp; Save</button>
      <div class="muted" style="margin-top:8px;">Images go to <code>static/images/portfolio/&lt;category&gt;</code>, videos to <code>static/videos/portfolio</code>.</div>
    </form>
    <div>
      <h2 style="margin:0 0 12px;">Current items</h2>
      <div id="items" class="grid"></div>
    </div>
  </div>

  <script>
    const catSelect = document.getElementById('category');
    const itemsEl = document.getElementById('items');

    async function loadData() {
      const res = await fetch('/api/data');
      const { categories, items } = await res.json();
      catSelect.innerHTML = categories.map(c => \`<option value="\${c.id}">\${c.name}</option>\`).join('');
      renderItems(items);
    }

    function renderItems(items) {
      if (!items.length) {
        itemsEl.innerHTML = '<p class="muted">No items yet.</p>';
        return;
      }
      itemsEl.innerHTML = items.map(item => {
        const media = item.video
          ? \`<video controls preload="metadata" src="\${item.video}"></video>\`
          : \`<img src="\${item.image}" alt="\${item.title}" loading="lazy" />\`;
        const tags = (item.tags || []).map(t => \`<span class="pill">\${t}</span>\`).join('');
        return \`
          <div class="card">
            \${media}
            <div class="content">
              <div style="font-weight:700;">\${item.title}</div>
              <div class="muted">\${item.category} • \${item.date || ''}</div>
              <div class="muted">\${item.description || ''}</div>
              <div class="tags">\${tags}</div>
            </div>
          </div>
        \`;
      }).join('');
    }

    document.getElementById('upload-form').addEventListener('submit', async (e) => {
      e.preventDefault();
      const form = e.currentTarget;
      const data = new FormData(form);
      const btn = form.querySelector('button');
      btn.disabled = true;
      btn.textContent = 'Uploading...';
      try {
        const res = await fetch('/api/upload', { method: 'POST', body: data });
        const json = await res.json();
        if (!json.ok) throw new Error(json.error || 'Upload failed');
        form.reset();
        await loadData();
        alert('Saved');
      } catch (err) {
        alert(err.message);
      } finally {
        btn.disabled = false;
        btn.textContent = 'Upload & Save';
      }
    });

    loadData();
  </script>
</body>
</html>`;
}

File: scripts/media_gui.py
#!/usr/bin/env python3
"""
Quick local GUI to add images/videos to the portfolio and keep JSON in sync.
Usage:
  python scripts/media_gui.py
Notes:
  - Saves images to static/images/portfolio/<category>/
  - Saves videos to static/videos/portfolio/
  - Updates src/lib/data/portfolio-data.json
  - Local-only; no auth. Close the window when done.
"""

from __future__ import annotations

import json
import shutil
import tkinter as tk
from datetime import date
from pathlib import Path
from tkinter import filedialog, messagebox

# Project paths
ROOT = Path(__file__).resolve().parent.parent
DATA_PATH = ROOT / "src" / "lib" / "data" / "portfolio-data.json"
IMAGE_ROOT = ROOT / "static" / "images" / "portfolio"
VIDEO_ROOT = ROOT / "static" / "videos" / "portfolio"

CATEGORIES = [
    ("3d-tours", "3D Tours"),
    ("interior", "Interior Photography"),
    ("exterior", "Exterior Photography"),
    ("drone", "Drone & Aerial"),
    ("virtual-staging", "Virtual Staging"),
    ("floorplans", "Floorplans (Cubicasa)"),
    ("portraits", "Portraits & Headshots"),
]

VIDEO_EXT = {".mp4", ".mov", ".m4v", ".avi", ".mkv", ".webm"}


def ensure_dirs() -> None:
    IMAGE_ROOT.mkdir(parents=True, exist_ok=True)
    VIDEO_ROOT.mkdir(parents=True, exist_ok=True)
    for cat, _ in CATEGORIES:
        (IMAGE_ROOT / cat).mkdir(parents=True, exist_ok=True)
    if not DATA_PATH.exists():
        DATA_PATH.parent.mkdir(parents=True, exist_ok=True)
        DATA_PATH.write_text("[]", encoding="utf-8")


def load_items() -> list[dict]:
    try:
        return json.loads(DATA_PATH.read_text(encoding="utf-8"))
    except Exception:
        return []


def save_items(items: list[dict]) -> None:
    DATA_PATH.write_text(json.dumps(items, indent=2), encoding="utf-8")


def slugify(name: str) -> str:
    cleaned = "".join(ch if ch.isalnum() else "-" for ch in name.lower())
    while "--" in cleaned:
        cleaned = cleaned.replace("--", "-")
    return cleaned.strip("-") or "media"


class MediaGui:
    def __init__(self) -> None:
        ensure_dirs()
        self.items = load_items()
        self.file_path: Path | None = None

        self.root = tk.Tk()
        self.root.title("Full Scope Media — Portfolio Manager")

        # Form fields
        self.category_var = tk.StringVar(value=CATEGORIES[0][0])
        self.title_var = tk.StringVar()
        self.desc_var = tk.StringVar()
        self.client_var = tk.StringVar()
        self.results_var = tk.StringVar()
        self.date_var = tk.StringVar(value=str(date.today()))
        self.tags_var = tk.StringVar()
        self.file_label_var = tk.StringVar(value="No file selected")
        self.status_var = tk.StringVar(value="Ready")

        self.build_ui()
        self.refresh_list()

    def build_ui(self) -> None:
        pad = {"padx": 8, "pady": 4, "sticky": "ew"}

        frame = tk.Frame(self.root)
        frame.grid(row=0, column=0, sticky="nsew")
        self.root.columnconfigure(0, weight=1)
        self.root.rowconfigure(0, weight=1)

        tk.Label(frame, text="Category").grid(row=0, column=0, **pad)
        tk.OptionMenu(frame, self.category_var, *[c[0] for c in CATEGORIES]).grid(
            row=0, column=1, **pad
        )

        tk.Label(frame, text="Title").grid(row=1, column=0, **pad)
        tk.Entry(frame, textvariable=self.title_var).grid(row=1, column=1, **pad)

        tk.Label(frame, text="Description").grid(row=2, column=0, **pad)
        tk.Entry(frame, textvariable=self.desc_var).grid(row=2, column=1, **pad)

        tk.Label(frame, text="Client").grid(row=3, column=0, **pad)
        tk.Entry(frame, textvariable=self.client_var).grid(row=3, column=1, **pad)

        tk.Label(frame, text="Results").grid(row=4, column=0, **pad)
        tk.Entry(frame, textvariable=self.results_var).grid(row=4, column=1, **pad)

        tk.Label(frame, text="Date (YYYY-MM-DD)").grid(row=5, column=0, **pad)
        tk.Entry(frame, textvariable=self.date_var).grid(row=5, column=1, **pad)

        tk.Label(frame, text="Tags (comma-separated)").grid(row=6, column=0, **pad)
        tk.Entry(frame, textvariable=self.tags_var).grid(row=6, column=1, **pad)

        tk.Button(frame, text="Choose file", command=self.choose_file).grid(
            row=7, column=0, **pad
        )
        tk.Label(frame, textvariable=self.file_label_var, anchor="w").grid(
            row=7, column=1, **pad
        )

        tk.Button(frame, text="Add to portfolio", command=self.add_item).grid(
            row=8, column=0, columnspan=2, padx=8, pady=8, sticky="ew"
        )

        self.listbox = tk.Listbox(frame, height=12)
        self.listbox.grid(row=9, column=0, columnspan=2, padx=8, pady=8, sticky="nsew")
        frame.rowconfigure(9, weight=1)

        tk.Label(frame, textvariable=self.status_var, anchor="w").grid(
            row=10, column=0, columnspan=2, padx=8, pady=4, sticky="ew"
        )

    def choose_file(self) -> None:
        path = filedialog.askopenfilename(
            title="Choose image or video",
            filetypes=[
                ("Media", "*.jpg *.jpeg *.png *.webp *.gif *.mp4 *.mov *.m4v *.avi *.mkv *.webm"),
                ("All files", "*.*"),
            ],
        )
        if path:
            self.file_path = Path(path)
            self.file_label_var.set(self.file_path.name)

    def add_item(self) -> None:
        if not self.file_path or not self.file_path.exists():
            messagebox.showerror("Missing file", "Please choose a file.")
            return

        cat = self.category_var.get()
        title = self.title_var.get().strip() or self.file_path.stem
        desc = self.desc_var.get().strip()
        client = self.client_var.get().strip()
        results = self.results_var.get().strip()
        tags = [t.strip() for t in self.tags_var.get().split(",") if t.strip()]
        date_str = self.date_var.get().strip() or str(date.today())

        ext = self.file_path.suffix.lower()
        is_video = ext in VIDEO_EXT
        dest_dir = VIDEO_ROOT if is_video else IMAGE_ROOT / cat
        dest_dir.mkdir(parents=True, exist_ok=True)

        safe_name = f"{slugify(title)}{ext}"
        dest_path = dest_dir / safe_name
        # Avoid overwrite collisions
        counter = 1
        while dest_path.exists():
            dest_path = dest_dir / f"{slugify(title)}-{counter}{ext}"
            counter += 1

        shutil.copy2(self.file_path, dest_path)

        items = load_items()
        next_id = max([int(i.get("id", 0)) for i in items] + [0]) + 1
        public_path = (
            f"/videos/portfolio/{dest_path.name}"
            if is_video
            else f"/images/portfolio/{cat}/{dest_path.name}"
        )

        new_item = {
            "id": next_id,
            "title": title,
            "category": cat,
            "description": desc,
            "image": None if is_video else public_path,
            "video": public_path if is_video else None,
            "tags": tags,
            "date": date_str,
            "client": client,
            "results": results,
        }

        items.append(new_item)
        save_items(items)
        self.items = items
        self.refresh_list()

        self.status_var.set(f"Added: {title}")
        messagebox.showinfo("Saved", f"Saved and copied to {dest_path}")

    def refresh_list(self) -> None:
        self.listbox.delete(0, tk.END)
        for item in sorted(self.items, key=lambda x: x.get("id", 0), reverse=True):
            self.listbox.insert(
                tk.END, f"#{item.get('id')} [{item.get('category')}] {item.get('title')}"
            )

    def run(self) -> None:
        self.root.mainloop()


if __name__ == "__main__":
    gui = MediaGui()
    gui.run()
